pennylane/interfaces/batch/torch.py

```diff
@@ -0,0 +1,236 @@
+# limitations under the License.
+"""
+This module contains functions for adding the Autograd interface
+to a PennyLane Device class.
+"""
+# pylint: disable=too-many-arguments,protected-access
+import inspect
+
+import numpy as np
+import torch
+
+import pennylane as qml
+
+
+def _compute_vjp(dy, jacs, device=None):
+    vjps = []
+
+    for d, jac in zip(dy, jacs):
+        if isinstance(jac, np.ndarray):
+            jac = torch.from_numpy(jac)
+
+        jac = torch.as_tensor(jac, device=device)
+        vjp = qml.gradients.compute_vjp(d, jac)
+        vjps.extend(vjp)
+
+    return vjps
+
+
+class ExecuteTapes(torch.autograd.Function):
+    """The signature of this ``torch.autograd.Function`` is designed to
+    workaround Torch restrictions.
+
+    In particular, ``torch.autograd.Function``:
+
+    - Cannot accept keyword arguments. As a result, we pass a dictionary
+      as the first argument ``kwargs``. This dictionary **must** contain:
+
+      * ``"tapes"``: the quantum tapes to batch evaluate
+      * ``"device"``: the device to use to evaluate the tapes
+      * ``execute_fn``: the execution function to use on forward passes
+      * ``"gradient_fn"``: the gradient transform function to use
+        for backward passes
+      * ``gradient_kwargs``: gradient keyword arguments to pass to the
+        gradient function
+      * ``max_diff``: the maximum order of derivatives to support
+
+    Further, note that the ``parameters`` argument is dependent on the
+    ``tapes``; this Function should always be called
+    with the parameters extracted directly from the tapes as follows:
+
+    >>> parameters = []
+    >>> [parameters.extend(t.get_parameters()) for t in tapes])
+    >>> kwargs = {"tapes": tapes, "device": device, "gradient_fn": gradient_fn, ...}
+    >>> ExecuteTapes.apply(kwargs, *parameters)
+
+    The private argument ``_n`` is used to track nesting of derivatives, for example
+    if the nth-order derivative is requested. Do not set this argument unless you
+    understand the consequences!
+    """
+
+    @staticmethod
+    def forward(ctx, kwargs, *parameters):  # pylint: disable=arguments-differ
+        """Implements the forward pass batch tape evaluation."""
+        ctx.tapes = kwargs["tapes"]
+        ctx.device = kwargs["device"]
+
+        ctx.execute_fn = kwargs["execute_fn"]
+        ctx.gradient_fn = kwargs["gradient_fn"]
+
+        ctx.gradient_kwargs = kwargs["gradient_kwargs"]
+        ctx.max_diff = kwargs["max_diff"]
+        ctx._n = kwargs.get("_n", 1)
+
+        with qml.tape.Unwrap(*ctx.tapes):
+            res, ctx.jacs = ctx.execute_fn(ctx.tapes, **ctx.gradient_kwargs)
+
+        # if any input tensor uses the GPU, the output should as well
+        ctx.torch_device = None
+
+        for p in parameters:
+            if isinstance(p, torch.Tensor):
+                if p.is_cuda:  # pragma: no cover
+                    cuda_device = p.get_device()
+                    break

```

Comment:
@glassnotes here is the GPU logic on the forward pass. Note that `torch.tensor(..., device=None)` seems to do nothing, which is nice.

Here, I am assuming that if _any_ parameter on _any_ tape is on the GPU, then... all outputs for every tape should be cast to GPU on output.

Reply:
With the fix above, everything gets put on the GPU. 

However issues arise later in the actual computation of the gradient with things not being on the same device. So in an example I ran, an error occurs in line 204 of `gradients/param_shift.py` where the results `res` are on the GPU, but the `gradient_coeffs` are not. (Actually this is the line you updated to use `tensordot` instead of `sum`.)

When doing `convert_like`, is there a way to pass the `gradient_coeffs` to the same device in a way that would be interface-agnostic? (maybe even as part of `cast_like`?)

Reply:
Oh good idea! Let me modify `convert_like`, and see if that fixes it for you.